### PR TITLE
[Reg2Mem] Add legacy pass wrapping Reg2Mem

### PR DESCRIPTION
**Pull request by Keenuts**
The SPIR-V backend will need to use Reg2Mem, hence this pass needs to be wrapped to be used with the legacy pass manager.

### CODE REVIEW

**Comment by Keenuts** [comment]
Hi! Here is some additional context on where we plan on using Reg2Mem pass if you need:
https://github.com/llvm/llvm-project/pull/111026

File: llvm/include/llvm/Transforms/Scalar/Reg2Mem.h [hunk 1]
```diff
@@ -22,6 +26,27 @@ class RegToMemPass : public PassInfoMixin<RegToMemPass> {
...
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.setPreservesAll();
+
+    AU.addPreserved<DominatorTreeWrapperPass>();
+    AU.addRequired<DominatorTreeWrapperPass>();
+
+    AU.addPreserved<LoopInfoWrapperPass>();
+    AU.addRequired<LoopInfoWrapperPass>();
+  }
```
Comment (nikic) [inline llvm/include/llvm/Transforms/Scalar/Reg2Mem.h:43]: Move this into the cpp file and remove related includes.

File: llvm/lib/Transforms/Scalar/Reg2Mem.cpp [hunk 2]
```diff
@@ -105,3 +105,29 @@ PreservedAnalyses RegToMemPass::run(Function &F, FunctionAnalysisManager &AM) {
   PA.preserve<LoopAnalysis>();
   return PA;
 }
+
+namespace llvm {
+void initializeRegToMemWrapperPassPass(PassRegistry &);
+} // namespace llvm
+
+INITIALIZE_PASS_BEGIN(RegToMemWrapperPass, "reg-to-mem", "", true, true)
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/Reg2Mem.cpp:113]: The pass is called reg2mem, not reg-to-mem.

**Comment by Keenuts** [comment]
Thanks! Moved to cpp, and renamed the pass to `reg2mem`.

File: llvm/lib/Transforms/Scalar/Reg2Mem.cpp [hunk 3]
```diff
@@ -105,3 +106,52 @@ PreservedAnalyses RegToMemPass::run(Function &F, FunctionAnalysisManager &AM) {
...
+RegToMemWrapperPass::RegToMemWrapperPass() : FunctionPass(ID) {}
+
+bool RegToMemWrapperPass::runOnFunction(Function &F) {
+  DominatorTree *DT = &getAnalysis<DominatorTreeWrapperPass>().getDomTree();
+  LoopInfo *LI = &getAnalysis<LoopInfoWrapperPass>().getLoopInfo();
+
+  unsigned N = SplitAllCriticalEdges(F, CriticalEdgeSplittingOptions(DT, LI));
+  bool Changed = runPass(F);
+  return N != 0 || Changed;
+}
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/Reg2Mem.cpp:153]: You can move these functions into the class declaration, not much point in moving them out of line if its the same file.

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/Scalar/Reg2Mem.cpp (hunk 3 above)
Comment (Keenuts) [reply llvm/lib/Transforms/Scalar/Reg2Mem.cpp]: Oh yeah of course.. Thanks!

**Comment by Keenuts** [comment]
Added the header functions to be able to instantiate it.

**Comment by Keenuts** [comment]
Back from holiday! Thanks for the review. Rebased on main, will wait for green CI and local tests & submit.

**Pull request merged**